Local storage

savvyserg opened the pull request:
### Features:
(**/src/localdb/main.rs**) 
- Local database websocket server binary crate (using existing WebsocketServer implementation)

(**/src/localdb/db/parsetools**) 
- Handy trait methods for parsing GlueSQL payloads

(**/src/lib/database**)
- LocalDbClient with managed connection pool (using deadpool)
- This is a replacement for the SimpleDbClient that used deadpool-postgres abstractions

(**/src/service/auth**, **/src/service/user**, **/src/service/admin**)
- Repository modules for each service to deal with the logic previously dealt with in PostgreSQL proc functions

(**/src/codegen/rust.rs**)
- Auto generated TryFrom<i32> implementations for every enum model using metaprogramming

(**/src/lib/id_gen.rs**)
- Snowflake id implementation for id generation (similar to Twitter’s implementation for generating tweet ids, but with service number instead of database shard)
- This effectively replaces the sequences previously created in the database itself

(**/README.md**)
- Revamped readme with a new graph, firecamp request collections (with comments on auth headers and such), and further explanations on the codebase modules

### Removals:
- Removed tokio-postgres and deadpool-postgres dependencies

(**/src/codegen**) 
- Codegen logic pertaining to procedural functions was removed since the database queries now use multiple steps of handcrafted logic (done by the repositories), and there are no procedural functions in GlueSQL
- Codegen for generating SQL models and schemas was removed
- Codegen for generating the database client struct and implementing functions that call database procs was removed

savvyserg committed:
local storage & create user table

- added gluesql dependency

- added localdb binary crate

- localdb now creates user table

savvyserg committed:
gluesql client struct

- added client submodule

- added localdb as a module of the main package

savvyserg committed:
database struct

- added database module with method to initialize local db

- added constants module with const array for sql create scripts

savvyserg committed:
user table indexes

- added index array to constants module

- added loop to create indexes in database init method

savvyserg committed:
user related tables

- added all user related tables, this means adding the tables "login_attempt",
"authorization_attempt", "reset_password_attempt", "recovery_question",
"recovery_question_data", "support_ticket" and "bad_request"

- added sql comments on key fields that must be handled by code due to lack
of gluesql features

savvyserg committed:
localdb module redesign

- database module is now the one that interfaces directly with gluesql

- client module aggregates the database module

- table and index creation scripts moved to a file named "create" in the "sql"
submodule

savvyserg committed:
database query result

- database query method now returns a result, and error handling is responsiblity
of the caller

- added comment about gluesql bug where create index statement will throw error
if index already exists even if using the "IF NOT EXISTS" directive

savvyserg committed:
localdb client read/write separation

- added read and write methods to client module

- added QueryChecker struct to database module

- QueryChecker is composed in Client struct

- added read and write query validation

savvyserg committed:
tokenized queries

- client's read and write methods now accept a query with "?" tokens and an
owned string vector

- added client module function to tokenize query

- added client module function to format token with literal quotes if non-numeric

- added validation for number of tokens

savvyserg committed:
repeatable tokens

- query placeholders now can be a question mark "?" followed by any sequence of
numeric digits

- placeholders with whatever number will be replaced by the corresponding token
in the order they are found in the query, e.g. the first placeholder in the
query will be replaced by the first token in the vector, wherever this same
placeholder is used again will be replaced by the same first token

savvyserg committed:
query tokenization fix

- fixed "unique_placeholders" function returning a vector with repeated placeholders

- fixed "tokenize_query" function incorrectly inserting tokens if some of the
placeholders used a subset of the same characters in the same order

savvyserg committed:
float token formatting fix

- fixed token formater recognizing as number a floating point character sequence
without an integer part

savvyserg committed:
better read-only assertion

- read/write agnostic statements are now allowed

- removed statements not applicable to PosgreSQL

- fixed CreateView statement being considered read-only

savvyserg committed:
readability refactor & redesign

- moved tokenization and statement check logic to it's own module "statements"

- renamed database module's query function to "exec"

- renamed all query string variables to "statements"

- improved some error messages

- changed code style to multiline when using too many combinators

savvyserg committed:
tokenization by index

- placeholders now must be comprised of a question mark "?" followed by the index
number of the token in the token vector

- added validation for placeholder index numbers

- optimized a bit by using more references and less clones

savvyserg committed:
idiomatic error handling

- added custom enum errors for all modules, wrapping the dependency error when
possible

- implemented the "From" trait for all custom errors

- added derived "Debug" trait for all custom errors

- removed all unnecessary explicit match statements

- separated "tokenizer" module validation into different functions

savvyserg committed:
tokenization fix

- fixed placeholder sorting

- added placeholder sorting custom error

- implemented stderr for TokenizerError

savvyserg committed:
localdb no more client abstraction

- removed client abstraction since the features it was planned for will not be
necessary

savvyserg committed:
localdb no more analyzer

- removed analyzer module, since it was a redundant statement analysis step and
localdb will no longer try to infer statement types before running the query,
this should be a significant optimization

savvyserg committed:
unique username constraint

- added unique constraint to username in user table

savvyserg committed:
database exported types & stderr trait

- added exported type aliases for critical internal gluesql types

- implemented stderr trait for DatabaseError

savvyserg committed:
db submodule

- moved all gluesql related logic to a db submodule in localdb

savvyserg committed:
database websocket server

- localdb is now serving queries through websocket in its binary crate

- added method module with handler

- added endpoint module

savvyserg committed:
config file as default environment values

- added default values for all services to config.json

- changed lib/config.rs to use config.json as default and overwrite them in
case they are passed as command line arguments

savvyserg committed:
rustfmt style

- changed code style to adhere to rust standards using rustfmt

savvyserg committed:
local database client

- added client to communicate with the database ran with the localdb binary

- added connection module and DbConnection struct opening, closing the websocket
connection as well as writing to sink and reading from stream

- added manager module with deadpool manager implementation to handle connections

- added stringable module with new "Stringable" trait that assures all arguments
sent to the client can be parsed as strings

- changed "connect_to_database" function to its own "connect" submodule

- removed deadpool_postgres dependency from the lib crate

- added deadpool dependency to the lib crate

- renamed client from "SimpleDbClient" to "LocalDbClient"

savvyserg committed:
localdb library crate

- added Cargo.toml file to src/localdb

- added localdb library crate as a dependency in the root Cargo.toml

- removed localdb reference by path from src/service/lib.rs

- changed localdb binary crate name to "db"

- formatted root Cargo.toml with rustfmt

savvyserg committed:
concise default env value implementation

- removed unnecessary if clauses

savvyserg committed:
better error handling

- added comprehensive error messages to config startup

savvyserg committed:
error naming standardization

- wrapped error propagating from submodule in error of the same name

savvyserg committed:
column name fix

- fixed the name of the "preferred_language" column in the "user" table

savvyserg committed:
boolean tokenization fix

- fixed boolean values being inserted to query string with quotes

savvyserg committed:
column constraint fix

- fixed columns "user_device_id" and "admin_device_id" from the "user" table
declared as NOT NULL

savvyserg committed:
login_attempt table constraint fix

- fixed some NULL fields declared as NOT NULL

savvyserg committed:
uuid serialization fix

- all uuid fields are now stored in GlueSQL as TEXT, this is because GlueSQL
internally stores uuids as u128, and since serde does not currenlty support
u128 serialization, a select to this field will cause the database server
tokio runtime to panic

savvyserg committed:
forced string flag

- tokenizer will now add quotes to token regardless of the formatting if the
string begins with the characters --forced-string

savvyserg committed:
user table fields fix

- fixed "reset_token_valid" field type

- changed "role" field name to "user_role" to fix GlueSQL mystical parsing bug

savvyserg committed:
gluesql payload parsetools

- added parsetools module with trait methods for parsing gluesql payload

- to access methods simply import like "use localdb::parsetools::*"

savvyserg committed:
auto trait implementation with metaprogramming

- added tryfrom<i32> trait implementation for all enum models

- added strum macros to enum models

savvyserg committed:
repositories

- added repositories for all routes and previously declared procs for
interfacing with the database

- subdivided the "method" submodules into "handler" and "repository"

savvyserg committed:
unused code warning removal

- added underscores to all implementations of proc functions that do not have a
route yet, to signal that it is intentional and improve readability

- added underscores to all errors only used in those routes that were not being
called

savvyserg committed:
fixes and readability pass

- fixed transactions possibly not closing

- fixed transactions committing even though an error occurred in one of the
statements

- fixed missing commas in queries

- fixed query whitespace

- added comments in the style of the previously used PostgreSQL procs

savvyserg committed:
list users response fix

- fixed list users response setting role number as "created_at" timestamp

savvyserg committed:
snowflake ids - id generation optimization

- ids are now generated with a snowflake approach, similar to Twitter's on tweet
ids, but with service id instead of shard id

- removed requests to database used to emulate sequences, which should optimize
id generation by an order of magnitude

- added "id_gen" module to "lib" with Snowflake and ConcurrentSnowflake structs

- added ConcurrentSnowflake to handlers

- added ConcurrentSnowflake as a parameter to repository functions

- added tests

savvyserg committed:
auto generated repository req & res models

- changed ProceduralFunction name to RepositoryFunction and removed its body

- all procedural model generation was kept, but procedural function generation
is no more

- changed all postgres and proc related names from "pg" to "repo"

savvyserg committed:
auto generated models derived traits

- added strum traits "EnumString" and "Display" for auto generated enums

- removed postgres related traits

savvyserg committed:
cleanup

- removed all metaprogramming functions that generated implementations to call
proc functions

- removed all metaprogramming functions that generated SQL code

- removed all SQL code

- removed test code in localdb's main function

- removed Stringable trait implementations for Vec<String> and Vec<i32>

- removed all references to deadpool-postgres and tokio-postgres

- removed all postgres related dependencies

savvyserg committed:
precision fix

- fixed Snowflake accepting service ids larger than 2 bits

- fixed error message

- added constants to replace magic numbers

savvyserg committed:
README Revamp

- added graph with authentication process

- added explanations for the auth header variables

- added firecamp request collections

- added instructions

JakkuSakura commented on README.md:
```diff
@@ -1,39 +1,114 @@
+
+Install [Firecamp]
+
+[Firecamp]: https://firecamp.io
```
Can firecamp be used out of the box? Especially with our login header and websocket-based API

savvyserg replied:
Firecamp is made for websockets and default value for headers come out of the box if you import the collections!

There are two complications with it though:
1- It does not yet save request bodies, that's why I specified the in the instructions for the user to get the bodies from the "environment" section, that I found is saved in the collection, so I put them there

2- It rounds large numbers (64 bit and upwards), this took me a very long time to debug, and is a major complication when using it with routes that require the user public_id, since the public_id displayed by firecamp is incorrect

Sadly I only found those complications out when I was almost finished, I ended up adding the collections because I thought it was better than not adding them (it provides some level of utility when starting from scratch, and it is really intuitive to establish websocket connections and send/receive messages)

But it is worth noting that it is a small project with only two contributors [(check it out)](https://github.com/firecamp-io/firecamp) so there is probably a lot of room for improvement

JakkuSakura replied:
> 2- It rounds large numbers (64 bit and upwards), this took me a very long time to debug, and is a major complication when using it with routes that require the user public_id, since the public_id displayed by firecamp is incorrect

It's the nature of javascript. The same happens when I was passing log id to FE in order to locate error messages. But I couldn't find the log id in log message. It turns out the i64 problem of JS. I haven't checked out firecamp yet but I guess it's JS

JakkuSakura replied:
Is configs for firecamp code-generated or hand-written?

JakkuSakura replied:
> 1- It does not yet save request bodies, that's why I specified the in the instructions for the user to get the bodies from the "environment" section, that I found is saved in the collection, so I put them there

Can you check this? The docs say it can save the request body.
https://firecamp.io/docs/clients/websocket/preparing-message-collection

JakkuSakura commented on src/lib/config.rs:
```diff
@@ -46,19 +53,72 @@ pub struct AppConfig {
-    config.app.priv_cert = args.priv_cert;
-    println!("App config {:#?}", config.app);
-    Ok(config)
+    let root: serde_json::Value = serde_json::from_str(&config)?;
```
can be simplfied to `let mut config: Config = serde_json::from_str(&config)?;`, after including `DbConfig` and `AppFileConfig` in Config

savvyserg replied:
I believe this simplification wouldn't work as it is since I removed the "name" field from `/etc/config.json`, but I could have misunderstood what you meant.

I could also put the "name" field back! 
I had removed it because it became redundant when I added separate configurations for each app, since the app name is the key of the entire app config object:
![current_app_config_example](https://user-images.githubusercontent.com/29493965/198750172-d5fe9814-3090-4a29-8532-dbc855f94c6b.png)

Adding the "name" field back to each app config could simplify the logic (AppFileConfig struct would no longer be needed), but the config would look like this:
![redundant_app_config_example](https://user-images.githubusercontent.com/29493965/198750182-e7655196-581e-42f5-ba4c-2b4bf1030d13.png)

JakkuSakura replied:
name field for each app doesn't matter. Previous the names are passed from main function. Actually I've implemented a better config in production code, and I'll post it here

JakkuSakura replied:
```rust

use crate::database::DatabaseConfig;
use crate::log::LogLevel;
use clap::Parser;
use eyre::*;
use serde::de::DeserializeOwned;
use serde::*;
use serde_json::Value;
use std::collections::HashMap;
use std::env::current_dir;
use std::fmt::Debug;
use std::path::PathBuf;

#[derive(Parser)]
#[clap(author, version, about, long_about = None)]
struct CliArgument {
    /// The path to config file
    #[clap(
        short,
        long,
        value_parser,
        value_name = "FILE",
        default_value = "etc/config.json",
        env = "CONFIG"
    )]
    config: PathBuf,
    #[clap(long, env = "LOG_LEVEL")]
    log_level: Option<LogLevel>,
    #[clap(long, env = "HOST")]
    /// The host to listen on
    host: Option<String>,
    #[clap(long, env = "PORT")]
    /// The port to listen on
    port: Option<u16>,
    #[clap(long, env = "DEBUG")]
    debug: bool,
    #[clap(long, env = "PUB_CERTS", value_delimiter = ',')]
    pub_certs: Option<Vec<String>>,
    #[clap(long, env = "PRIV_CERT")]
    priv_cert: Option<String>,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Config<App> {
    pub app_db: DatabaseConfig,
    pub auth_db: DatabaseConfig,
    #[serde(skip)]
    pub app: AppConfig<App>,
    #[serde(flatten)]
    pub apps: HashMap<String, AppConfig<Value>>,
}
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AppConfig<App> {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub log_level: LogLevel,
    #[serde(default)]
    pub host: String,
    #[serde(default)]
    pub port: u16,
    #[serde(default)]
    pub pub_certs: Option<Vec<String>>,
    #[serde(default)]
    pub priv_cert: Option<String>,
    #[serde(default)]
    pub debug: bool,
    #[serde(skip)]
    pub header_only: bool,
    #[serde(flatten)]
    pub extra: App,
}
pub fn load_config<App: DeserializeOwned + Debug + Default>(
    service_name: String,
) -> Result<Config<App>> {
    let args: CliArgument = CliArgument::parse();

    println!("Working directory {}", current_dir()?.display());
    println!("Loading config from {}", args.config.display());
    let config = std::fs::read_to_string(&args.config)?;
    let mut config: Config<App> = serde_json::from_str(&config)?;
    if let Some(a) = config.apps.get(&service_name) {
        config.app = serde_json::from_value(serde_json::to_value(a)?)?;
    }

    // merge configs
    if let Some(log_level) = args.log_level {
        config.app.log_level = log_level;
    }

    if let Some(host) = args.host {
        config.app.host = host;
    }
    if let Some(port) = args.port {
        config.app.port = port;
    }
    config.app.name = service_name;
    if let Some(pub_certs) = args.pub_certs {
        config.app.pub_certs = Some(pub_certs);
    }
    if let Some(priv_cert) = args.priv_cert {
        config.app.priv_cert = Some(priv_cert);
    }
    if args.debug {
        config.app.debug = true;
    }
    println!("App config {:#?}", config.app);
    Ok(config)
}

```

savvyserg replied:
thanks man, already fixed and pushed to local-storage branch

I added the better version of the config.rs following this file, I removed the generics, and had to change a couple of things in `ws/server`

JakkuSakura commented on src/lib/config.rs:
```diff
@@ -46,19 +53,72 @@ pub struct AppConfig {
+    Ok(Config { db: db, app: app })
+}
+
+#[derive(Debug)]
+pub enum ConfigError {
+    DefaultDbConfigNotFoundError(String),
+    DefaultAppConfigNotFoundError(String),
```
maybe use the `eyre` library if we don't care about the type of errors. e.g. `eyre::bail!("DbConfigNotFoundError")` in place of `ConfigError::DefaultDbConfigNotFoundError("xx")`.
If we do care about error types(like in a formal library for external audience), use `thiserror` instead

savvyserg replied:
Had to add some match statements to make it work, but now it doesn't have custom errors, what do you think?

Before:
![config_before](https://user-images.githubusercontent.com/29493965/198753326-8928c669-d3b2-42dd-aa90-b20c95d04641.png)

After:
![config_after](https://user-images.githubusercontent.com/29493965/198753333-44fbdbcf-9a2e-4ed5-be39-8c19baa85c49.png)

savvyserg replied:
> If we do care about error types(like in a formal library for external audience), use thiserror instead

Do you mean I should change the custom errors in all files to thiserror? I could do that if it sounds better!

Sorry about that btw, I added quite a lot of custom errors because it's the only way I could find to wrap any kind of error in order to use the question mark operator

I tried to prioritize making the code readable and idiomatic, but my lack of experience with Rust might have gotten the best of me :p

JakkuSakura replied:
It's not a must but you'll find out converting to thiserror is fairly simple, saving a lot of code. It's a better way to write custom errors

savvyserg replied:
Converted all custom errors to this error, already pushed to local-storage branch

Thanks!

JakkuSakura commented on src/lib/database/stringable.rs:
```diff
@@ -0,0 +1,92 @@
+use hex::encode;
+
+pub trait Stringable {
+    fn stringify(&self) -> Result<String, StringifyError>;
```
In any case, `Stringable::stringify` fails?

savvyserg replied:
You're right, stringify is currently infallible and no longer should return a Result, it was a leftover signature from when I tried to make it parse more complex types

Thank you! 

Already fixed and pushed to the local-storage branch

JakkuSakura commented on src/lib/id_gen.rs:
```diff
@@ -0,0 +1,226 @@
+const MAX_17_BITS: u32 = 131071;
+const MAX_2_BITS: u16 = 3;
+
+pub struct ConcurrentSnowflake {
```
nice work with snowflake algorithm

savvyserg replied:
Thank you! 

It currently occasionally wastes a millisecond if under a lot of stress, I’m planning on fixing that by adding more state to the Snowflake struct

JakkuSakura commented on src/lib/toolbox.rs:
```diff
@@ -155,23 +155,6 @@ impl Toolbox {
                     return;
                 }
-
-                Err(err0) if err0.downcast_ref::<tokio_postgres::Error>().is_some() => {
```
we used to return certain kinds of error from postgresql, and passthrough to the client. I'm not sure this pass through can be achieved in the local storage

savvyserg replied:
I know, the ability for the services to return an error to the client that originates in the localdb websocket server is currently pending

I'm aware that removing and not replacing this code is not good, but I was at a loss when I last tried to implement it, and this would have been dead code if I had left it

That said, I could still try to implement it for this change!

JakkuSakura replied:
It's fine here, as the database doesn't have a procedural function implementation

JakkuSakura commented on src/localdb/db/sql/create.rs:
```diff
@@ -0,0 +1,116 @@
+pub const TABLES: [&str; 8] = [
```
better use `include_str!("../../../db/tbl.sql")`, or something similar. This provides us with IDE support. It's also easier to modify with external tools

savvyserg replied:
This makes a lot of sense, thank you 

Already fixed and pushed to local-storage branch, the init method in /src/localdb/database now looks like this:
![current_database_init](https://user-images.githubusercontent.com/29493965/198764598-7aa6416d-faa5-4976-9756-b0ea17d27457.png)

JakkuSakura commented on src/localdb/parsetools/value.rs:
```diff
@@ -0,0 +1,222 @@
+
+use super::Value;
+
+pub trait ValueParser {
+    fn try_i8(&self) -> Result<i8, ParseValueError>;
+    fn possible_i8(&self) -> Result<Option<i8>, ParseValueError>;
+    fn try_i16(&self) -> Result<i16, ParseValueError>;
+    fn possible_i16(&self) -> Result<Option<i16>, ParseValueError>;
+    fn try_i32(&self) -> Result<i32, ParseValueError>;
+    fn possible_i32(&self) -> Result<Option<i32>, ParseValueError>;
+    fn try_i64(&self) -> Result<i64, ParseValueError>;
+    fn possible_i64(&self) -> Result<Option<i64>, ParseValueError>;
+    fn try_i128(&self) -> Result<i128, ParseValueError>;
+    fn possible_i128(&self) -> Result<Option<i128>, ParseValueError>;
+    fn try_f64(&self) -> Result<f64, ParseValueError>;
+    fn possible_f64(&self) -> Result<Option<f64>, ParseValueError>;
+    fn try_bool(&self) -> Result<bool, ParseValueError>;
+    fn possible_bool(&self) -> Result<Option<bool>, ParseValueError>;
+    fn try_string(&self) -> Result<String, ParseValueError>;
+    fn possible_string(&self) -> Result<Option<String>, ParseValueError>;
+    fn try_bytea(&self) -> Result<Vec<u8>, ParseValueError>;
+    fn possible_bytea(&self) -> Result<Option<Vec<u8>>, ParseValueError>;
+    fn try_uuid(&self) -> Result<Uuid, ParseValueError>;
+    fn possible_uuid(&self) -> Result<Option<Uuid>, ParseValueError>;
+}
+
```
Possible to simplify `ValueParser` with `Parsable`?

savvyserg replied:
I'm sorry, I'm a bit confused, do you mean I should use the [Parsable crate](https://docs.rs/parsable/latest/parsable/#)? Or maybe that I should rename the Trait? Or combine the traits used for parsing the GlueSQL payload?

JakkuSakura replied:
First rename the trait,  then study postgres' FromSQL and ToSQL traits

savvyserg replied:
Renames pushed to local-storage branch, will study after I finish the other changes

Thanks for the tips!

JakkuSakura commented on src/service/auth/method/repository.rs:
```diff
@@ -0,0 +1,979 @@
+use lib::id_gen::{ConcurrentSnowflake, ConcurrentSnowflakeError};
+use uuid::Uuid;
+
+use gen::database::{
+    FunAuthAuthenticateReq, FunAuthAuthenticateResp, FunAuthAuthenticateRespRow,
+    FunAuthAuthorizeReq, FunAuthAuthorizeResp, FunAuthAuthorizeRespRow,
+    FunAuthBasicAuthenticateReq, FunAuthBasicAuthenticateResp, FunAuthBasicAuthenticateRespRow,
+    FunAuthChangePasswordReq, FunAuthChangePasswordResp, FunAuthChangePasswordRespRow,
+    FunAuthGetPasswordSaltReq, FunAuthGetPasswordSaltResp, FunAuthGetPasswordSaltRespRow,
+    FunAuthGetRecoveryQuestionsReq, FunAuthGetRecoveryQuestionsResp,
+    FunAuthGetRecoveryQuestionsRespRow, FunAuthResetPasswordReq, FunAuthResetPasswordResp,
+    FunAuthResetPasswordRespRow, FunAuthSetRecoveryQuestionsReq, FunAuthSetRecoveryQuestionsResp,
+    FunAuthSetRecoveryQuestionsRespRow, FunAuthSetTokenReq, FunAuthSetTokenResp,
+    FunAuthSetTokenRespRow, FunAuthSignupReq, FunAuthSignupResp, FunAuthSignupRespRow,
+    FunGetRecoveryQuestionDataReq, FunGetRecoveryQuestionDataResp,
+    FunGetRecoveryQuestionDataRespRow, FunSubmitRecoveryAnswersReq, FunSubmitRecoveryAnswersResp,
+    FunSubmitRecoveryAnswersRespRow,
+};
```
`use gen::database::*`

savvyserg replied:
Thank you, fixed and pushed to local-storage branch

JakkuSakura commented on src/service/auth/method/repository.rs:
```diff
@@ -0,0 +1,979 @@
+
+use gen::model::{EnumRecoveryQuestionCategory, EnumRole, EnumService};
+
+pub async fn fun_auth_signup(
```
DALs mixed with rust and SQL? Works with this demo, but not suitable in places where transactions are needed.
I understand that procedural functions are rarely implemented in small databases. Some devs write DAL with ORM, while we use plpgsql with code-generated bindings.

savvyserg replied:
I'm not sure I understand the alternative, I implemented the repository functions to mimic the previous procedural function behaviors as closely as possible using GlueSQL behind a websocket server, this meant using Rust to throw some Exceptions and to do assertions that were previously done in pure SQL

Also noteworthy are some hacky stuff like multi-request transactions (GlueSQL does not yet rollback a transaction on COMMIT if one of the statements threw an Exception, so I had to check transaction status and manually trigger the rollback when applicable), I'm not proud of it, but the challenge was to change the database into something that wouldn't need a container, and preferably rust based, and I did my best to do that

I totally get if you guys don't think this a good solution, though

That being said, if we decide that I am to continue working on this change the way it is, could you elaborate a little more on how you imagine this being done?

JakkuSakura replied:
The alternative means that
1. you put db function inside GlueSQL server
2. you call these db functions remotely(these handles/stubs are code generated)
The current solution is that:
1. you implement a GlueSQL server
2. you run db functions in app server, which issues SQL to GlueSQL server

JakkuSakura replied:
In one day we'll finally roll our own database for all this kind of stuff, JK

savvyserg replied:
> In one day we'll finally roll our own database for all this kind of stuff, JK

Lol, you guys are insane, I love it

> The alternative means that
> 1. you put db function inside GlueSQL server
> 2. you call these db functions remotely(these handles/stubs are code generated)
> The current solution is that:
> 3. you implement a GlueSQL server
> 4. you run db functions in app server, which issues SQL to GlueSQL server

I think I get it, and It's pretty cool, but if I put these functions inside the server, even if I can expose some way to call them remotely, they will still be a mix of SQL and Rust since GlueSQL's SQL does not have features like if/else clauses, functions, throw Exception, or declare variable!

Or are you talking about implementing procedural functions in GlueSQL? 

I'm getting a feeling this change will get out of control quickly, lol

JakkuSakura replied:
> I think I get it, and It's pretty cool, but if I put these functions inside the server, even if I can expose some way to call them remotely, they will still be a mix of SQL and Rust since GlueSQL's SQL does not have features like if/else clauses, functions, throw Exception, or declare variable!

The mixture of SQL and Rust is unavoidable unless we implement procedural functions in GlueSQL somehow. Since this is a demo, we can have them mixed for now.

savvyserg committed:
stringify fallibility fix

- the "stringify" trait method from the "Stringable" trait is now infallible and
no longer returns a Result

- removed "Stringable" custom errors

- the "parse_req_params" function from the database/client module is also now
infallible

savvyserg committed:
import succinctness

- changed explicit repository imports from gen::database to gen::database::*

savvyserg committed:
IDE support for gluesql table SQL

- moved SQL statements to create table to /db/tbl.sql

- adapted /src/localdb/database init function

savvyserg committed:
better config

- changed config according to more streamlined logic according to reference

- removed config custom errors

- changed config.json anatomy to have basic db info, db server and all services
in one level of depth

savvyserg committed:
parsetools rename

savvyserg committed:
thiserror

- all custom errors now use thiserror
